tools/canonicalize-proto-file/Main.hs

```diff
@@ -11,28 +10,110 @@
+  , type (<?>) (..)
+  , unwrapRecord
+  )
+
+import Prelude hiding (FilePath)
+
+import Proto3.Suite.DotProto.AST 
+  ( DotProto (..)
+  , DotProtoDefinition (..)
+  , DotProtoEnumPart (..)
+  , DotProtoEnumValue 
+  , DotProtoField (..)
+  , DotProtoIdentifier (..)
+  , DotProtoImport (..)
+  , DotProtoMessagePart (..)
+  , DotProtoOption (..)
+  , DotProtoPackageSpec (..)
+  , DotProtoServicePart (..)
+  , DotProtoReservedField (..)
+  , DotProtoType (..)
+  , DotProtoValue (..)
+  , Path (..)
+  , RPCMethod (..)
+  )
+import Proto3.Suite.DotProto.Generate (readDotProtoWithContext)
+import Proto3.Suite.DotProto.Rendering (defRenderingOptions, toProtoFile)
+import Proto3.Wire.Types (FieldNumber (..))
+
+import Turtle (FilePath)
+
+--------------------------------------------------------------------------------
 
 data Args w = Args
   { includeDir :: w ::: [FilePath] <?> "Path to search for included .proto files (can be repeated, and paths will be searched in order; the current directory is used if this option is not provided)"
   , proto      :: w ::: FilePath   <?> "Path to input .proto file"
   } deriving Generic
+
 instance ParseRecord (Args Wrapped)
+
 deriving instance Show (Args Unwrapped)
 
+-- | Check if the given values is some given bounds (inclusive).
+between :: (Ord a, Enum a) => a -> (a, a) -> Bool
+between x i = fst i <= x && x <= snd i
+
+-- | @('isOverlappingIntervals' a b :: 'Bool')@ is a relation between two 
+-- interval-like types @a@ and @b@ that is 'True' when the two intervals are 
+-- overlapping or touching over a subinterval of @a@ and @b@.
+isOverlappingIntervals :: (Ord a, Enum a) => (a, a) -> (a, a) -> Bool
+isOverlappingIntervals i1@(x, y) i2@(u, v) = x `between` i2 || y `between` i2 || u `between` i1 || v `between` i1
+
+-- | @('joinIntervals' a b :: 'Maybe' (a, a))@ will join the two given intervals 
+-- @a@ and @b@ into a larger interval if 
+--
+-- @
+-- ('isOverlappingIntervals' a b '==' 'True')
+-- @
+--
+-- Otherwise, returns 'Nothing'.
+joinIntervals :: (Ord a, Enum a) => (a, a) -> (a, a) -> Maybe (a, a)
+joinIntervals a b = do 
+  guard (isOverlappingIntervals a b)
+  pure (min (fst a) (fst b), max (snd a) (snd b))
+
+-- | Normalizes a list of intervals. 
+--
+-- 1. Filters out any "invalid" intervals from the resulting list, i.e. any 
+--    intervals @(x, y) :: (a, a)@ with @x > y@. 
+--
+-- 2. Additionally, the resulting list will have all overlapping intervals 
+--    merged into a larger interval that covers each overlapping interval.
+normalizeIntervals :: (Ord a, Enum a) => [(a, a)] -> [(a, a)]
+normalizeIntervals = mergeIntervals . filter \(x, y) -> x <= y
+
+-- | Returns a sorted list that contains all intervals from the minimal set of 
+-- intervals to represent the given list of intervals. 
+--
+-- "Merges" overlapping intervals in a list of intervals. Think disjunctive 
+-- normal form.
+mergeIntervals :: (Ord a, Enum a) => [(a, a)] -> [(a, a)]
+mergeIntervals = foldr step [] . sort
```

Comment:
This implementation looks correct, but could you provide a unit test of some sort?  For example, you could restrict to, say, [0 .. 99], then exhaustively iterate through the two-digit numbers, checking that each number that is in none of the input intervals is in none of the result intervals, and that each number that is in at least one of the input intervals is in exactly one of the output intervals.  That would avoid creating a reference implementation as such--it would just check the desired properties, and with only 100 numbers to check, it should finish promptly.

Reply:
Yeah, I can do. I'll move the code for merging intervals into `Proto3.Suite.DotProto.Internal` and import it into the `canonicalize-proto-file/Main.hs` file. This way both the executable and the test suite have access to the code.